### PR TITLE
implement response transmittable for async enrollments expirations

### PR DESCRIPTION
**Pull request by saipraveen18**
# PR Checklist

Please check if your PR fulfills the following requirements
- [X] The title follows our [guidelines](https://github.com/ideacrew/enroll/blob/trunk/CONTRIBUTING.md#commit
)
- [X] Tests for the changes have been added (for bugfixes/features)

# PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature (requires Feature flag)
- [ ] Data fix or migration (inert code, no impact until run)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Dependency updates (e.g., add a new gem or updated version)

# What is the ticket # detailing the issue?

Ticket: [IVL Product Development - 186580335](https://www.pivotaltracker.com/story/show/186580335)

# A brief description of the changes

Current behavior: Currently, on the first of the year, the enrollments expire both synchronously and asynchronously based on the resource registry configuration.

New behavior: Transmittable models are created asynchronously when enrollments expire, specifically when the resource registry `:async_expire_and_begin_coverages` is enabled for tracking purposes.

# Feature Flag

For all new feature development, a feature flag is required to control the exposure of the feature to our end users. A feature flag needs a corresponding environment variable that is used to initialize the state of the flag. Please share the name of the environment variable below that would enable/disable the feature and which client(s) it applies to.

Variable name: `:async_expire_and_begin_coverages`

- [ ] DC
- [ ] ME

# Additional Context
Include any additional context that may be relevant to the peer review process.

# AppScan CodeSweep Failure
In the event of a failed check on the AppScan CodeSweep step of our GitHub Actions workflow, please review the False Positive protocol outlined here: appscan_codesweep/CODESWEEP_FALSE_POSITIVES_README.MD

Add all required notes to this section if the failure is a suspected false positive.

### CODE REVIEW

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 1]
```diff
@@ -4,38 +4,193 @@ module Operations
...
+
+        if transmission.success?
+          status_result = update_status("Transmittable::Transmission found with given global ID: #{transmission_gid}",
+                                        :succeeded,
+                                        { transmission: transmission.value! })
+
+          return status_result if status_result.failure?
+          Success(transmission.value!)
+        else
+          add_errors(:invalid_transmission_on_enrollment_expire,
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:114]: @saipraveen18 I think we should simply return a failure monad if we are unable to find the Transmission. This operation can be called with invalid params and we should not add errors to the job. This specific step should be split into 2 steps.
- Find the transmission by global_id
- Updating the Status of Transmission

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 2]
```diff
@@ -4,38 +4,193 @@ module Operations
...
+          msg = "Failed to expire enrollment hbx id #{hbx_enrollment.hbx_id} - #{hbx_enrollment.kind} is not a valid IVL enrollment kind"
+          Failure(msg)
+        else
+          Success(hbx_enrollment)
+        end
       end
 
-      def find_enrollment(enrollment_hbx_id)
-        Operations::HbxEnrollments::Find.new.call({hbx_id: enrollment_hbx_id})
+      def find_and_update_request_transmission(transmission_gid, job)
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:103]: I think we should simply return a failure monad if we are unable to find the Transmission. This operation can be called with invalid params and we should not add errors to the job. This specific step should be split into 2 steps.
- Find the transmission by global_id
- Updating the Status of Transmission

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 3]
```diff
@@ -4,38 +4,193 @@ module Operations
...
+                                        { job: job })
+          return status_result if status_result.failure?
+          msg = "No Transmittable::Transmission found with given global ID: #{transmission_gid}"
+          Failure(msg)
+        end
       end
 
-      def expire_enrollment(enrollment)
-        return Failure("Failed to expire enrollment hbx id #{enrollment.hbx_id} - #{enrollment.kind} is not a valid IVL enrollment kind") unless enrollment.is_ivl_by_kind?
+      def find_and_update_request_transaction(transaction_gid, job, request_transmission)
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:126]: Same as Transmission.

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 4]
```diff
@@ -4,38 +4,193 @@ module Operations
...
+                  publish_on: Date.today,
+                  started_at: DateTime.now,
+                  event: 'received',
+                  transaction_id: hbx_enrollment.hbx_id,
+                  state_key: :received
+                })
+      end
 
-        Success("Successfully expired enrollment hbx id #{enrollment.hbx_id}")
+      def expire_enrollment
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:178]: Updating the Status of Response Transmission and transaction on successful enrollment transition can be extracted into a separate step.

**Review by saikumar9 — CHANGES_REQUESTED** [review]

File: app/domain/operations/hbx_enrollments/expire.rb (hunk 2 above)
Comment (mdkaraman) [reply app/domain/operations/hbx_enrollments/expire.rb:103]: I agree with @saikumar9.  If we are checking for success/failure of finding the transmission inside this step, it seems more appropriate to move that into its own step in the `call` method up top.  Same goes for the `find_and_update_request_transaction` method and anywhere else we are checking for success/failure inside a method that returns a separate success/failure.

**Review by mdkaraman — CHANGES_REQUESTED** [review]
I tested expirations end to end on my local machine and they working as expected, so no issues in terms of functionality.  However, I agree with Sai's request for refactor.

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 5]
```diff
@@ -4,38 +4,193 @@ module Operations
...
+        Success({
+                  transmission: response_transmission,
+                  subject: hbx_enrollment,
+                  key: :hbx_enrollment_expiration_response,
+                  title: "Enrollment expiration response transaction for #{hbx_enrollment.hbx_id}.",
+                  description: "Transaction response to expire enrollment with hbx id: #{hbx_enrollment.hbx_id}.",
+                  publish_on: Date.today,
+                  started_at: DateTime.now,
+                  event: 'received',
+                  transaction_id: hbx_enrollment.hbx_id,
```
Comment (mdkaraman) [inline app/domain/operations/hbx_enrollments/expire.rb:173]: I think this needs to be `correlation_id`, even though the field on the transaction model is `transaction_id`.  Check out: https://github.com/ideacrew/enroll/blob/8ce54c1e297e31903338b2dfb90f77fd3e58371d/app/domain/operations/transmittable/create_transaction.rb#L39

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 6]
```diff
@@ -4,38 +4,150 @@ module Operations
...
+        request_transmission    = yield find_transmission_by_global_id(values[:transmittable_identifiers][:transmission_gid])
+        _transmission_result    = yield update_status("Transmittable::Transmission found with given global ID: #{values[:transmittable_identifiers][:transmission_gid]}",
+                                                      :succeeded,
+                                                      { transmission: request_transmission})
+        request_transaction     = yield find_transaction_by_global_id(values[:transmittable_identifiers][:transaction_gid])
+        _transaction_result     = yield update_status("Transmittable::Transaction found with given global ID: #{values[:transmittable_identifiers][:transaction_gid]}",
+                                                      :succeeded,
+                                                      { transaction: request_transaction})
+        transmission_params     = yield construct_response_transmission_params(job)
+        @response_transmission  = yield create_response_transmission(transmission_params,
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:38]: @saipraveen18 As discussed let us add errors only on Job.

File: app/domain/operations/hbx_enrollments/expire.rb [hunk 7]
```diff
@@ -4,38 +4,150 @@ module Operations
...
+                                                                     { job: job,
+                                                                       transmission: request_transmission,
+                                                                       transaction: request_transaction})
+        transaction_params      = yield construct_response_transaction_params
+        @response_transaction   = yield create_response_transaction(transaction_params,
```
Comment (saikumar9) [inline app/domain/operations/hbx_enrollments/expire.rb:43]: @saipraveen18 As discussed let us add errors only on Job.

**Review by saikumar9 — CHANGES_REQUESTED** [review]
A couple of other nitpicks.
- Refactor code in 'app/domain/operations/hbx_enrollments/publish_expiration_event.rb' to not pass `enrollment_hbx_id` in the payload
- Refactor code in 'app/event_source/subscribers/individual/enrollments/expire_coverages_subscriber.rb' to log `enrollment_gid` instead of `enrollment_hbx_id` in the subscriber logger.

**Review by mdkaraman — DISMISSED** [review]
Tested this locally end to end and confirmed that expirations are working as expected.  Also checked on the transmittable data, which is being populated accurately as far as I can tell.  I am approving because I see no other issues from my perspective.  Nice work!

**Review by saikumar9 — APPROVED** [review]

**Review by mdkaraman — APPROVED** [review]

**Pull request merged**